modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2324,37 +2324,48 @@ def fold(self, axis, func, new_columns=None):
             The axis to apply over.
         func : callable
             The function to apply.
+        new_index : list-like, optional
+            The index of the result.
         new_columns : list-like, optional
             The columns of the result.
-            Must be the same length as the columns' length of `self`.
-            The column labels of `self` may change during an operation so
-            we may want to pass the new column labels in (e.g., see `cat.codes`).
 
         Returns
         -------
         PandasDataframe
             A new dataframe.
-
-        Notes
-        -----
-        The data shape is not changed (length and width of the table).
         """
+        copy_lengths = True
+        copy_widths = True
+        if new_index is not None:
+            if self.has_materialized_index:
+                if len(self.index) != len(new_index):
+                    copy_lengths = False
+            else:
+                copy_lengths = False
         if new_columns is not None:
             if self.has_materialized_columns:
-                assert len(self.columns) == len(
-                    new_columns
-                ), "The length of `new_columns` doesn't match the columns' length of `self`"
-            self.set_columns_cache(new_columns)
+                if len(self.columns) != len(new_columns):
+                    copy_widths = False
+            else:
+                copy_widths = False
 
         new_partitions = self._partition_mgr_cls.map_axis_partitions(
             axis, self._partitions, func, keep_partitioning=True
         )
         return self.__constructor__(
             new_partitions,
-            self.copy_index_cache(copy_lengths=True),
-            self.copy_columns_cache(copy_lengths=True),
-            self._row_lengths_cache,
-            self._column_widths_cache,
+            (
+                self.copy_index_cache(copy_lengths=copy_lengths)
```

Comment:
It seems that we cannot copy the cache index in general now.

Reply:
That's why it is under the condition.